### ISSUE TITLE
Import NeTEx_wsProducer-Document.wsdl

### ISSUE THREAD

**Issue by procfxgen** [NeTEx-CEN/NeTEx#72]
Hi,
  I try to import "NeTEx_wsProducer-Document.wsdl" from VS2017 in order to generate C# Classes.

 I have some errors while importing service:

Severity	Code	Description	Project	File	Line	Suppression State
Warning		Custom tool warning: Cannot import wsdl:binding
Detail: There was an error importing a wsdl:portType that the wsdl:binding is dependent on.
XPath to wsdl:portType: //wsdl:definitions[@targetNamespace='http://www.siri.org.uk/siriWS']/wsdl:portType[@name='NetexProducerDocPort']

XPath to Error Source: //wsdl:definitions[@targetNamespace='http://www.siri.org.uk/siriWS']/wsdl:binding[@name='NetexProducerDocBinding']	ExportNetex	C:\aaa\projects\SAEIV_sol\NETEX\ExportNetex\ExportNetex\Connected Services\ServiceReference1\Reference.svcmap	1	

Thanks.

**Comment by skinkie** [comment]
In the last have year I have been working with some partners that were trying the same. I have received word that generating the classes from the Microsoft tools does not result in much success. There are some other paid products that seem to work for C#. It is still in our best interest to have our XSD / WSDL compatible with other products than xjb (java).

**Comment by nick-knowles** [comment]
Agreed -  the question is how to arrive at a list of revisions needed to fix?

**Comment by skinkie** [comment]
> Agreed - the question is how to arrive at a list of revisions needed to fix?

Some changes here resulted in XSD.exe to actually do something. But within DATA4PT we might get Microsoft on board for some actual development.

https://github.com/FlippieTrippie/NeTEx/commits/XSD.exe

**Comment by acranton** [comment]
Hi,

I'm trying to create C# classes using xsd.exe and even after adding in the comments suggested in [https://github.com/FlippieTrippie/NeTEx/commits/XSD.exe](https://github.com/FlippieTrippie/NeTEx/commits/XSD.exe) I still get a circular reference error: 

'SiteComponentGroup' from targetNamespace='http://www.netex.org.uk/netex' has invalid definition: Circular group reference.

Do you know how to get around this or if there is anywhere I can get a copy of the C# classes for this?

Thanks,
Alex

**Comment by skinkie** [comment]
@acranton sadly we still don't have a magic bullet for C#.

**Comment by acranton** [comment]
Thanks for letting me know.  I support and develop a large PT database for a couple of local authorities which currently exports to TransXChange and just starting to plan how best to incorporate a NetEx export.  Guess I will need to try and manually create the class definitions based on the schema if there is no easy alternative.

**Comment by skinkie** [comment]
> Thanks for letting me know. I support and develop a large PT database for a couple of local authorities which currently exports to TransXChange and just starting to plan how best to incorporate a NetEx export. Guess I will need to try and manually create the class definitions based on the schema if there is no easy alternative.

There are some appoaches. It seems that some commercial XML packages are able to generate classes that do work in C# but it might be that you would need them at runtime, which is bad in my perspective. The alternative is to use an existing NeTEx file, and generate classes from that. And you could switch to a language that would support this better...

**Comment by acranton** [comment]
Yeah I had a quick look at some of those but definitely don't want to be shipping an extra third party .dll file, I don't like that idea either.  I think ill try using an XML file as a starting point, I just find the arrays/special types tend to require quite a bit of manual attention.  If you could point me in the direction of a fairly comprehensive example NetEx xml file I could use for the import that would be really helpful.

Its tricky to introduce another language at this stage as the core application is bespoke, 20+ years old and written in Delphi (which has terrible serialization support) but is millions of lines of code so cant easily be migrated, but we have quite a few C# based modules/services/APIs which plug into it that I was planning to expand for this.  I also haven't done any Java for a very long time so would be an added learning curve..

**Comment by skinkie** [comment]
Don't blame Delphi it works fine ;) I have been looking for projects that do XSD to 'other languages', I didn't find any of them out of the box supporting our NeTEx XSD. So we could do some work ourselves, but I have also been reaching out to many parties (including Microsoft) to solve the XSD to code issues.

**Comment by acranton** [comment]
Haha, I have been working in Delphi alongside C# since forever, it definitely still gets the job done but the syntactic sugar in C# does speed things up quite a lot! Plus hard to get support on Delphi online these days for those less standard type issues...

Thats good to hear, please do let me know if you manage to find a way to convert or hear of anyone who is happy to share the classes.  Im probably not going to be taking this forwards immediately but its definitely on the todo list - if I end up creating the classes in C# ill probably make them available publicly (although haven't completely ruled out using Delphi just yet as the TXC code is all in there at the moment and can potentially re-use some of the data objects, although was going to move these to C# if the serialization was straightforward...)

**Comment by skinkie** [comment]
@acranton Ulf shared a possible solution. I have asked to post it here.

**Comment by acranton** [comment]
Thanks - any help gratefully received!

**Comment by skinkie** [comment]
@acranton  (and others in this issue) would any of you be available to test classes? I currently have only one person in my "favor" list that is actually willing to test the results for C#, I would hope I could speed up the process if some other persons would be able to test what we currently have, to establish a baseline.

**Comment by acranton** [comment]
@skinkie yes, no problem, happy to help in any way to have class support for C#

**Comment by skinkie** [comment]
@acranton could you send me an e-mail at stefan a opengeo d nl?

**Comment by budul100** [comment]
> 
> 
> > Hi,
> > I'm trying to create C# classes using xsd.exe and even after adding in the comments suggested in https://github.com/FlippieTrippie/NeTEx/commits/XSD.exe I still get a circular reference error:
> > 'SiteComponentGroup' from targetNamespace='http://www.netex.org.uk/netex' has invalid definition: Circular group reference.
> > Do you know how to get around this or if there is anywhere I can get a copy of the C# classes for this?
> > Thanks, Alex
> 
> Hi Alex,
> 
> It is an old issue but at least I have found a "kind of" workaround. I have commented out the following parts to get around this exception (actually it would be better to determine the circle reference but I haven't found the root cause yet):
> 
> In _netex_ifopt_site_version.xsd_:
> 
> ```
> 
> 
> ```
> 
> In _netex_organisation_version.xsd_:
> 
> ```
> 
> ```
> 
> And eventually in _netex_journeyPattern_version.xsd_:
> 
> ```
> 
> ```
> 
> Then at least, I was able to create a C# model of NetEx part 1.
> 
> Regards, budul

Actually, I have found a solution now which solves the problems above completely. Instead of using wsd.exe, I have created the classes by using the [XmlSchemaClassGenerator](https://github.com/mganss/XmlSchemaClassGenerator). By using this tool I was able to avoid any circular reference issues and it can furthermore create separate files for each class.

Together with the recommendation from @skinkie you should have a very fast and reliable solution.

**Comment by skinkie** [comment]
@budul100 indead, that tool is a life saver if you compare it to the standard C# stuff (and even the proprietary onces).

**Comment by ue71603** [comment]
@skinkie  what do we do with this issue? Create a FAQ on github and refer to this one, but closed? Or do we have work here to do?

**Comment by skinkie** [comment]
Documentation, FAQ etc.
